typeclasses/maps.py

```diff
@@ -0,0 +1,252 @@
+        '''
+        player = self.current_character.player
+        is_builder = player.locks.check_lockstring(player, 'dummy:perm(Builders)')
+
+        min_x = max_x = 0
+        min_y = max_y = 0
+        for room_info in self.db.map_data.values():
+            x = room_info['x']
+            y = room_info['y']
+            if x < min_x: min_x = x
+            if y < min_y: min_y = y
+            if x > max_x: max_x = x
+            if y > max_y: max_y = y
+
+        max_x += 1
+        max_y += 1
+
+        width = max_x - min_x
+        height = max_y - min_y
+
+        alt_grid = None
+        if is_builder:
+            alt_grid = []
+            for y in range(height):
+                alt_grid.append([])
+                for x in range(width):
+                    alt_grid[y].append(' ' * CELL_WIDTH)
+            for room, room_info in self.db.map_data.items():
+                alt_grid[room_info['y'] - min_y][room_info['x'] - min_x] = str(room.id).center(CELL_WIDTH)
+
+
+        map_grid = []
+        for y in range(height):
+            map_grid.append([])
+            for x in range(width):
+                map_grid[y].append(' ' * CELL_WIDTH)
+        for room, room_info in self.db.map_data.items():
+            map_grid[room_info['y'] - min_y][room_info['x'] - min_x] = self.tile_for_room(room)
+
+        if show_player and not self.db.off_map:
+            x = self.db.x - min_x
+            y = self.db.y - min_y
+            try:
+                tile = list(map_grid[y][x])
+            except IndexError: pass
+            else:
+                tile[int(len(tile) / 2)] = '*'
+                map_grid[y][x] = ''.join(tile)
+
+        if alt_grid:
+            for i, row in reversed(list(enumerate(alt_grid))):
+                map_grid.insert(i, row)
+        return map_grid
+
+    def tile_for_room(self, room):
+        '''
+        Get the tile for the provided `room`,
+        or `?!?` if no tile could be found.
+        '''
+        tile = None
+        try:
+            tile = room.map_tile
+        except AttributeError: pass
+        else:
+            tile = room.map_tile(self)
+
+        if tile is None:
+            tile = room.db.map_tile
+        if tile is None:
+            tile = TILES.get(room.terrain, '?!?')
+
+        return tile
+
+
+    def return_appearance(self, looker):
+        '''
+        Return the map data nicely wrapped up in a box:
+        ┌─── MAP ───┐
+        │ map  here │
+        └───────────┘
+        '''
+        map_grid = self.draw_map()
+        width = CELL_WIDTH * len(map_grid[0]) + PADDING * 2
+
+        map_page = []
+        title_line = title = u' MAP '
+        # str.center is dumb. Work around that.
+        spacer = u'─'
+        i = 0
+        while len(title_line) < width:
+            if i % 2:
+                title_line += spacer
+            else:
+                title_line = spacer + title_line
+            i += 1
```

Comment:
Hm. I could do `title.center(width, '#')` and replace all of the `#` characters with the box drawing character.